### PR TITLE
Add API endpoint for shelving

### PR DESCRIPTION
**Pull request by jcoyne**
## Why was this change made?

So that common-accessioning doesn't need to be aware of how to shelve something. It can just ensure the process is triggered. 

## Was the API documentation (openapi.json) updated?
Yes

### CODE REVIEW

File: app/controllers/shelves_controller.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        errors: [
+          { title: 'Invalid item type', detail: "A Dor::Item is required but you provided '#{@item.class}'" }
+        ]
+      }, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+  def druid_tools
```
Comment (justinlittman) [inline app/controllers/shelves_controller.rb]: Is this used?

File: app/services/digital_stacks_service.rb [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  # Delete files from stacks that have change type 'deleted', 'copydeleted', or 'modified'
+  # @param [Pathname] stacks_object_pathname the stacks location of the digital object
+  # @param [Moab::FileGroupDifference] content_diff the content file version differences report
+  def self.remove_from_stacks(stacks_object_pathname, content_diff)
+    %i[deleted copydeleted modified].each do |change_type|
+      subset = content_diff.subset(change_type) # {Moab::FileGroupDifferenceSubset}
+      subset.files.each do |moab_file| # {Moab::FileInstanceDifference}
+        moab_signature = moab_file.signatures.first # {Moab::FileSignature}
+        file_pathname = stacks_object_pathname.join(moab_file.basis_path)
+        delete_file(file_pathname, moab_signature)
```
Comment (justinlittman) [inline app/services/digital_stacks_service.rb]: Ignoring when delete_file doesn't succeed (returns false).

File: app/services/digital_stacks_service.rb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  # @param [Moab::FileGroupDifference] content_diff the content file version differences report
+  def self.rename_in_stacks(stacks_object_pathname, content_diff)
+    subset = content_diff.subset(:renamed) # {Moab::FileGroupDifferenceSubset
+
+    # 1st Pass - rename files from original name to checksum-based name
+    subset.files.each do |moab_file| # {Moab::FileInstanceDifference}
+      moab_signature = moab_file.signatures.first # {Moab::FileSignature}
+      original_pathname = stacks_object_pathname.join(moab_file.basis_path)
+      temp_pathname = stacks_object_pathname.join(moab_signature.checksums.values.last)
+      rename_file(original_pathname, temp_pathname, moab_signature)
```
Comment (justinlittman) [inline app/services/digital_stacks_service.rb]: Ignores when rename fails.

File: app/services/digital_stacks_service.rb [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    return false if workspace_content_pathname.nil?
+
+    %i[added copyadded modified].each do |change_type|
+      subset = content_diff.subset(change_type) # {Moab::FileGroupDifferenceSubset
+      subset.files.each do |moab_file| # {Moab::FileInstanceDifference}
+        moab_signature = moab_file.signatures.last # {Moab::FileSignature}
+        filename = change_type == :modified ? moab_file.basis_path : moab_file.other_path
+        workspace_pathname = workspace_content_pathname.join(filename)
+        stacks_pathname = stacks_object_pathname.join(filename)
+        copy_file(workspace_pathname, stacks_pathname, moab_signature)
```
Comment (justinlittman) [inline app/services/digital_stacks_service.rb]: Ignores when copy file fails.

File: app/services/shelving_service.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+# Push file changes for shelve-able files into stacks
+class ShelvingService
+  def self.shelve(work)
```
Comment (justinlittman) [inline app/services/shelving_service.rb]: Why calling "work" instead of "item"?

File: app/services/digital_stacks_service.rb (hunk 2 above)
Comment (jcoyne) [reply app/services/digital_stacks_service.rb]: I'm just moving these from common-accessioning, so I'd like to take a bye on this.

File: app/services/digital_stacks_service.rb (hunk 3 above)
Comment (jcoyne) [reply app/services/digital_stacks_service.rb]: This is how it is in common-accessioning.

File: app/services/digital_stacks_service.rb (hunk 4 above)
Comment (jcoyne) [reply app/services/digital_stacks_service.rb]: This is how it is in common-accessioning.

File: app/services/shelving_service.rb (hunk 5 above)
Comment (jcoyne) [reply app/services/shelving_service.rb]: This is how it is in common-accessioning.

File: app/controllers/shelves_controller.rb (hunk 1 above)
Comment (jcoyne) [reply app/controllers/shelves_controller.rb]: nope.

File: app/services/digital_stacks_service.rb (hunk 2 above)
Comment (jcoyne) [reply app/services/digital_stacks_service.rb]: https://github.com/sul-dlss/common-accessioning/blob/master/lib/digital_stacks_service.rb#L13

File: app/services/digital_stacks_service.rb (hunk 3 above)
Comment (jcoyne) [reply app/services/digital_stacks_service.rb]: https://github.com/sul-dlss/common-accessioning/blob/master/lib/digital_stacks_service.rb#L45

File: app/services/digital_stacks_service.rb (hunk 4 above)
Comment (jcoyne) [reply app/services/digital_stacks_service.rb]: https://github.com/sul-dlss/common-accessioning/blob/master/lib/digital_stacks_service.rb#L88

File: app/services/shelving_service.rb (hunk 5 above)
Comment (jcoyne) [reply app/services/shelving_service.rb]: https://github.com/sul-dlss/common-accessioning/blob/master/lib/shelving_service.rb#L5

**Pull request merged**